### DISCUSSION TITLE
Extracting a json subtree for processing (without preprocessing with jq)

### DISCUSSION THREAD

**Discussion by smammy** [johnkerl/miller#1512] [Q&A] [answered]
Say I have a JSON document like:
```json
{
    "a": [
        {"x":1, "y":2, "z":3},
        {"x":4, "y":5, "z":6},
        {"x":7, "y":8, "z":9}
    ],
    "b":[
        {"i":1, "j":2},
        {"i":3, "j":4},
        {"i":5, "j":6}
    ]
}
```
and I'm only interested in the records inside "a". What I usually do is extract the relevant subtree with jq and pass that to Miller:
```sh
cat test.json | jq .a | mlr --j2p cat # or whatever
```
Output:
```
x y z
1 2 3
4 5 6
7 8 9
```
This is fine I guess, but is there a pure-Miller way to do it? Bonus points if it's concise.

**Comment by johnkerl** [comment, marked as answer]
Hi @smammy !

`jq` is perhaps a bit more natural here, as it puts maps (objects) and arrays (lists) on equal footing. So `jq .a test.json` is fine whether the value at the `"a"` key is a scalar, list, or object.

Miller is a record processor -- it takes as Millerable JSON input (see also [here](https://miller.readthedocs.io/en/latest/file-formats/#json):

* A single object
* A list of objects

This means that the most intuitive idea does not work:

```
$ mlr -j --from test.json put '$* = $a'
Cannot assign non-map to existing map; got array.
```

A second optional (not bonus-level concise IMO) is

```
$ mlr -j --from test.json put -q '@records = $a; end { emit @records }'
[
{
  "records": [
    {
      "x": 1,
      "y": 2,
      "z": 3
    },
    {
      "x": 4,
      "y": 5,
      "z": 6
    },
    {
      "x": 7,
      "y": 8,
      "z": 9
    }
  ]
}
]
```

A third option is to rely on the fact that `print` produces JSON output:

```
$ mlr -j --from test.json put -q 'print $a'
[
  {
    "x": 1,
    "y": 2,
    "z": 3
  },
  {
    "x": 4,
    "y": 5,
    "z": 6
  },
  {
    "x": 7,
    "y": 8,
    "z": 9
  }
]
[
]
```

This is nice and concise IMO. However, there's that dangling `[]` at the end which is a bug. Probably the same bug as this:

```
$ mlr --json cat nonesuch
[
]
mlr: open nonesuch: no such file or directory.
```

**Reply by smammy** [reply]
Thanks, this makes a lot of sense!

**Comment by johnkerl** [comment]
#1513 was more subtle than I thought -- so @smammy I'm going with suggestion two above as my answer to your question.
